### ISSUE TITLE
Recovery "what to send section"?  / Pseudo code issue 

### ISSUE THREAD

**Issue by larseggert** [quicwg/base-drafts#1724]
i think there is an issue in the pseudo code (but I may be wrong - I find the code much more difficult to follow than an actual textual explanation.)

In `OnAckReceived`, there is the following call sequence:
```
       DetectLostPackets(ack.largest_acked_packet)
       SetLossDetectionTimer()
```

`DetectLostPackets` calls `OnPacketsLost`, which reduces `in_flight` potentially to zero. Then in `SetLossDetectionTimer`, the timer is cancelled via `loss_detection_timer.cancel()`.

This causes no retransmissions to happen, at least when (blindly) implementing the pseudo code.

Related to that, it is not clear where exactly the pseudo code expects actual retransmissions to happen. There is code in `OnLossDetectionTimeout` (`RetransmitAllUnackedHandshakeData`, `SendOnePacket`, `SendTwoPackets`) but what those do is not explained. There is also a comment in `DetectLostPackets` ("Inform the congestion controller...") but it remains unclear when and how that would lead to retransmissions.

IMO, if we rely on pseudo code for this document, it needs a lot more work to make sure it actually does what is intended. (My preference would be to instead describe the concepts in text.)

**Comment by mirjak** [comment]
I believe this is general issue with TLP which shows up as you are sending only with a low constant rate. We observed a similar problem in TCP...

**Comment by mirjak** [comment]
However, +1 to more text explanation!

**Comment by larseggert** [comment]
The pseudo code shouldn't have "general issues" :-)

**Comment by ianswett** [comment]
There needs to be some more clarity here, and I believe the core issue is this document is trying to describe both how to quickly detect packets as lost and congestion control, but is avoiding what send a bit too much.

Given QUIC doesn't retransmit packets, and instead retransmits stream data/etc, just because a packet is lost doesn't mean it's payload needs to be retransmitted, and if it does, it may be bundled differently from the original transmission.

Would a section on "What to send" be helpful?  Or would it be better to work that into the existing text and/or pseudocode?

**Comment by martinthomson** [comment]
Was this addressed in #1939?  The comment there says "part of" this only.  @ianswett, what is missing?

**Comment by ianswett** [comment]
Between that PR an some subsequent cleanups, I think we're in good shape.
